### PR TITLE
Add standalone example of using the SQL frontend

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/10855

## Rationale for this change

The current `rewrite_expr.rs` example has three distinct examples:
1. Using DataFusion LogicalPlanning without execution
2. Implementing an `AnalyzerRule`
3. Implementing an `OptimizerRule`

Each I think should be its own example to make them easier to find and more full featured

## What changes are included in this PR?

1. Add example `sql_frontend.rs`
2. Improve documentation of TableSource trait

## Are these changes tested?
Yes by CI

### CODE REVIEW

**Comment by alamb** [comment]
@andygrove  I wonder if you might have some time to review this PR

**Comment by alamb** [comment]
I am working on getting the CI to run successfully on this PR, see https://github.com/apache/datafusion/pull/11173

**Comment by alamb** [comment]
🤔  this might also make https://github.com/apache/datafusion/blob/main/datafusion/sql/examples/sql.rs redundant

File: datafusion-examples/examples/sql_frontend.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    let logical_plan = sql_to_rel.sql_statement_to_plan(statements[0].clone())?;
+    println!(
+        "Unoptimized Logical Plan:\n\n{}\n",
+        logical_plan.display_indent()
+    );
+
+    // The initial LogicalPlan is a mechanical translation from the parsed SQL
+    // and often can not run without the Analyzer passes.
+    //
+    // In this example, `person.age` is actually a different data type (Int32)
```
Comment (mustafasrepo) [inline datafusion-examples/examples/sql_frontend.rs:75]: At the source schema data type defined as `Uint8`

File: datafusion-examples/examples/sql_frontend.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    println!(
+        "Analyzed Logical Plan:\n\n{}\n",
+        analyzed_plan.display_indent()
+    );
+
+    // As we can see, the Analyzer added a CAST so the types are the same
+    // (Int64). However, this plan is not as efficient as it could be, as it
+    // will require casting *each row* of the input to UInt64 before comparison
+    // to 21 and 32. To optimize this query's performance, it is  better to cast
+    // the constants once at plan time to Int32.
```
Comment (mustafasrepo) [inline datafusion-examples/examples/sql_frontend.rs:102]: I guess this should be `to Int64`

**Review by mustafasrepo — APPROVED** [review]
Thanks @alamb for this PR.

File: datafusion-examples/examples/sql_frontend.rs (hunk 2 above)
Comment (alamb) [reply datafusion-examples/examples/sql_frontend.rs]: Good catch -- fixed in 6152fd675

File: datafusion-examples/examples/sql_frontend.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    // To invoke SqlToRel we must provide it schema and function information
+    // via an object that implements the `ContextProvider` trait
+    //
+    // Projection: person.name
+    //   Filter: person.age BETWEEN Int64(21) AND Int64(32)
+    //     TableScan: person
+    let context_provider = MyContextProvider::default();
+    let sql_to_rel = SqlToRel::new(&context_provider);
+    let logical_plan = sql_to_rel.sql_statement_to_plan(statements[0].clone())?;
+    println!(
```
Comment (comphead) [inline datafusion-examples/examples/sql_frontend.rs:67]: maybe we can use `assert!` here to let the user know the expected output?

File: datafusion-examples/examples/sql_frontend.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    // Projection: person.name
+    //   Filter: CAST(person.age AS Int64) BETWEEN Int64(21) AND Int64(32)
+    //     TableScan: person
+    let config = OptimizerContext::default().with_skip_failing_rules(false);
+    let analyzed_plan = Analyzer::new().execute_and_check(
+        logical_plan,
+        config.options(),
+        observe_analyzer,
+    )?;
+    println!(
```
Comment (comphead) [inline datafusion-examples/examples/sql_frontend.rs:93]: assert?

File: datafusion-examples/examples/sql_frontend.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    // TableScan: person projection=[name], full_filters=[person.age >= UInt8(21), person.age <= UInt8(32)]
+    //
+    // The optimizer did several things to this plan:
+    // 1. Removed casts from person.age as we described above
+    // 2. Converted BETWEEN to two single columns  inequalities (which are typically faster to execute)
+    // 3. Pushed the projection of `name` down to the scan (so the scan only returns that column)
+    // 4. Pushed the filter into the scan
+    let optimized_plan =
+        Optimizer::new().optimize(analyzed_plan, &config, observe_optimizer)?;
+    println!(
```
Comment (comphead) [inline datafusion-examples/examples/sql_frontend.rs:118]: assert?

File: datafusion-examples/examples/sql_frontend.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    fn get_table_source(&self, name: TableReference) -> Result<Arc<dyn TableSource>> {
+        if name.table() == "person" {
+            Ok(Arc::new(MyTableSource {
+                schema: Arc::new(Schema::new(vec![
+                    Field::new("name", DataType::Utf8, false),
+                    Field::new("age", DataType::UInt8, false),
+                ])),
+            }))
+        } else {
+            plan_err!("table not found")
```
Comment (comphead) [inline datafusion-examples/examples/sql_frontend.rs:149]:
```suggestion
            plan_err!("Table {} not found", name.table())
```

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

File: datafusion-examples/examples/sql_frontend.rs (hunk 3 above)
Comment (alamb) [reply datafusion-examples/examples/sql_frontend.rs:67]: that is a good idea -- I will do so (the existing examples are inconsistent with that).

Comment (comphead) [reply datafusion-examples/examples/sql_frontend.rs]: Filed https://github.com/apache/datafusion/issues/11230

**Comment by alamb** [comment]
Thanks everyone for the comments and feedback

**Pull request merged**